### PR TITLE
Add ES2015 (ES6) generator

### PR DESCRIPTION
**Pull request by darrylyeo**
I didn't like how the generated ES5 code looked in my ES2015 project, so I thought I'd make myself a generator.

### CODE REVIEW

File: lib/generate.js [hunk 1]
```diff
@@ -135,6 +135,33 @@ generate.javascript.builtinPostprocessors = {
...
+    return "// Generated automatically by nearley\n" +
+        "// http://github.com/Hardmath123/nearley\n" +
+        "{\n" +
+        "const id = x => x[0];\n" +
+        parser.body.join('\n') +
+        "const grammar = {\n" +
+        "    Lexer: " + parser.config.lexer + ",\n" +
+        "    ParserRules: " + serializeRules(parser.rules, generate.es2015.builtinPostprocessors) + ",\n" +
+        "    ParserStart: " + JSON.stringify(parser.start) + "\n" +
+        "}\n" +
```
Comment (deltaidea) [inline lib/generate.js]: Missing `;`

File: lib/generate.js [hunk 2]
```diff
@@ -135,6 +135,33 @@ generate.javascript.builtinPostprocessors = {
...
+        "    module.exports = grammar;\n" +
+        "} else {\n" +
+        "    window." + exportName + " = grammar;\n" +
+        "}\n" +
+        "}\n";
+};
+generate.es2015.builtinPostprocessors = {
+    "joiner": "d => d.join('')",
+    "arrconcat": "d => [d[0]].concat(d[1])",
+    "arrpush": "d => d[0].concat([d[1]])",
```
Comment (deltaidea) [inline lib/generate.js]: Could've done `([el, arr]) => [el].concat(arr)` and `([arr, el]) => arr.concat([el])`

File: lib/generate.js (hunk 1 above)
Comment (darrylyeo) [inline lib/generate.js]: Heh. Side effect of copying from the regular JS generator (see line 120).

If it's all right with you, I'd love to make another PR to make the coding style consistent.

File: lib/generate.js (hunk 2 above)
Comment (darrylyeo) [inline lib/generate.js]: Ooh, good idea.

File: lib/generate.js (hunk 1 above)
Comment (deltaidea) [inline lib/generate.js]: I'm just a fellow contributor, but sure! There're no other PRs to the codebase at the moment, so no potential conflicts.

It may be a good idea to make an issue first and discuss code styleguide properly, because there's none and the style is all over the spectrum. If you'd champion that issue, I'd be a happy man. I currently only have time to slowly inch forward with the docs over at #254.

**Comment by tjvr** [comment]
Thanks for getting involved in Nearley! :-)

I'm afraid I don't think we're going to merge this PR. The generated code isn't meant to be read, and it has no benefit apart from aesthetics to keep around an ES6 version

Regarding updating the code style of the project: we'd like to stick with ES5/UMD for compatibility, performance, and to avoid transpilers such as Babel.

Personally I'd prefer a style that dropped semicolons (like how [moo](tjvr/moo) is written), but it's not particularly worth arguing about: there are more important issues!

We at least want to stick with ES5 for the node core (ie. `nearley.js`). The tests already target Node 6 / ES6, and so it's plausible that the surrounding tooling—which only targets node—could use ES6. But I'm not sure @hardmath123 is ready for that yet :-)

**Comment by independentgeorge** [comment]
@tjvr So there is currently no way to use the generated grammar and `nearley.js` from inside an ES6 project, and there are no plans to add support for it?

**Comment by tjvr** [comment]
> there is currently no way to use the generated grammar and nearley.js from inside an ES6 project

That's not the case at all! ES6 is, of course, backwards-compatible with ES5; so you can use the generated grammar and nearley.js as normal. :-)

**Comment by independentgeorge** [comment]
By ES6 I meant the use of ES modules. I'm using [rollup](https://github.com/rollup/rollup) in my project and the UMD wrappers are getting in the way of `import`s.

I'm asking about support for importing `nearley.js` directly as an ES6 module, and specifying the output format of the generated grammar to ES6 modules so they could be imported as well.

I know this is not related to the PR, and I apologise if I'm asking in the wrong place, but you did mention some of what I'm talking about.

**Comment by tjvr** [comment]
I see. That probably should be a separate issue, yes! But since we’re here, I’ll reply.

I don’t think we’re going to support native `import`, at least not for a good while. Surely Rollup must provide backwards-compatibility with UMD modules... if not, then I don’t see it as nearley’s fault :-)
